enum Type {
  INT8    = 1;
  INT16   = 2;
  INT32   = 3;
  INT64   = 4;

  FLOAT32 = 5;
  FLOAT64 = 6;
}

enum Unit {
  SECOND = 1;
  CYCLE  = 2;
  VOLT   = 3;
}

enum Format {
  DENSE  = 1;
  SPARSE = 2;
}

enum Order  {
  TRACE  = 1;
  SAMPLE = 2;
}

message Value {
  oneof data {
    uint32 i8  = 1;
    uint32 i16 = 2;
    uint32 i32 = 3;
    uint32 i64 = 4;

    float  f32 = 5;
    double f64 = 6;
  }
}

message Sample {
  required Value   value = 1;
  optional uint32  index = 2;
}

message Meta {
  required bytes     key = 1;
  required bytes   value = 2;
  optional uint32  index = 3;
}

message Axis {
  required Type     type = 1;
  required Unit     unit = 2;

  optional float     min = 3;
  optional float     max = 4;
  optional float  offset = 5;
}

message Container {
  optional uint32 version_major =  1 [default = 0];
  optional uint32 version_minor =  2 [default = 1];

  required uint32 trace_count   =  3;
  required uint32 trace_length  =  4;
  optional uint32 trace_offset  =  5;

  required Format format        =  6;
  required Order  order         =  7;
  repeated Meta   meta          =  8;

  required Axis   axis_x        =  9;
  repeated Axis   axis_y        = 10;

  repeated Sample samples       = 11;
}
